src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,433 @@
+
+/*
+ * Copyright (C) cuiweixie
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_util.h"
+#include "ngx_http_lua_semaphore.h"
+#include "ngx_http_lua_contentby.h"
+
+int ngx_http_lua_ffi_set_semaphore_threshold(ngx_int_t threshold);
+static ngx_http_lua_semaphore_t *ngx_http_lua_alloc_sem(ngx_log_t *log);
+void ngx_http_lua_cleanup_sem(void *data);
+static void ngx_http_lua_free_sem(ngx_http_lua_semaphore_t *sem, ngx_log_t *log);
+static ngx_int_t ngx_http_lua_semaphore_resume(ngx_http_request_t *r);
+int ngx_http_lua_ffi_sem_new(ngx_http_lua_semaphore_t **psem, int n, char **errstr);
+int ngx_http_lua_ffi_sem_post(ngx_http_lua_semaphore_t *sem, int n, char **errstr);
+int ngx_http_lua_ffi_sem_wait(ngx_http_request_t *r, ngx_http_lua_semaphore_t *sem,
+                              int time, u_char *errstr, size_t *errlen);
+static void ngx_http_lua_ffi_semaphore_cleanup(void *data);
+void ngx_http_lua_ffi_semaphore_handler(ngx_event_t *ev);
+static void ngx_http_lua_sem_timeout_handler(ngx_event_t *ev);
+void ngx_http_lua_ffi_sem_gc(ngx_http_lua_semaphore_t *sem);
+
+static ngx_http_lua_semaphore_t *ngx_http_lua_sem_free_list = NULL;
+static ngx_int_t ngx_http_lua_semaphore_threshold = 0;
+static ngx_int_t ngx_http_lua_sem_free_count = 0;
```

Comment:
Can we avoid these global C variables above? They *may* cause problems during HUP reload. Better hook them up in the `ngx_http_lua_main_conf_t` construct.